session.go

```diff
@@ -551,41 +551,61 @@ func Connect(ctx context.Context, opts ...ConnectOption) (Session, error) {
 	}
 
 	sess := tunnel_client.NewReconnectingSession(logger, rawDialer, stateChanges, reconnect)
-
-	select {
-	case <-ctx.Done():
-		return nil, ctx.Err()
-	case err := <-stateChanges:
-		if err != nil {
+	// allow consumers to .Close() the session before a successful connect
+	session.setInner(&sessionInner{
+		Session: sess,
+	})
+
+	// performs one "pump" of the session update channel
+	// returns true if there are more updates to handle
+	runSessionHandlers := func() (bool, error) {
+		select {
+		case <-ctx.Done():
```

Comment:
i think we might want to run the disconnect handler here too, since otherwise consumers have no way of knowing their session died due to a done context after the initial connect succeeds.